overlord/state.go

```diff
@@ -20,25 +20,49 @@
 package overlord
 
 import (
+	"encoding/json"
+	"fmt"
 	"io"
 )
 
 // State represents a snapshot of the system state.
-type State struct{}
+type State struct {
+	entries map[string]json.RawMessage
+}
+
+// NewState returns an empty system state.
+func NewState() *State {
+	return &State{
+		entries: make(map[string]json.RawMessage),
+	}
+}
 
 // Get unmarshals the stored value associated with the provided key
 // into the value parameter.
 func (s *State) Get(key string, value interface{}) {
+	err := json.Unmarshal(s.entries[key], value)
+	if err != nil {
+		panic(fmt.Errorf("internal error: could not retrieve and unmarshal state entry %q: %v", key, err))
```

Comment:
I suggest adding an error result here, with two possible outcomes: a clear JSON unmarshal issue, or ErrNoState, returned when the key is missing altogether.